### ISSUE TITLE
Make auto-formatting more convenient with pre-commit

### ISSUE THREAD

**Issue by tobiasraabe** [hmgaudecker/econ-project-templates#48]
Hi,

I had two issues with the ``format_python_files.py`` approach which let me to change this in [my template](https://github.com/tobiasraabe/cookiecutter-research-template).

1. It is annoying to execute the formatting when you already agreed that files should look this way.
2. isort does not really work when used inside tox as it misrepresents first, second, third party applications which caused tests to fail.

Therefore, I now use a different approach using [pre-commit](https://pre-commit.com). The tool allows you to define several git hooks which will be execute when trying to commit. In case any of the defined hooks catches an error, the commit is aborted.

All you need is a [config file](https://github.com/tobiasraabe/cookiecutter-research-template/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/.pre-commit-config.yaml) and you need to install all packages under ``repo: local``. Then, type ``pre-commit install`` and the hooks will run before each commit. The posted config runs black, black on documentation, flake8 and doc8. Furthermore, isort is replaced with reorder_python_imports which is simpler and easier to implement.

There are a lot of other hooks available under https://pre-commit.com/hooks.html as yaml checks, syntax upgrades, etc..

Best

**Comment by hmgaudecker** [comment]
Sounds great, thanks for the pointer! Will implement it eventually.

**Comment by julienschat** [comment]
I tested pre-commit in one installation of the template with the [configuration file](https://github.com/tobiasraabe/cookiecutter-research-template/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/.pre-commit-config.yaml) and it worked perfectly. Do you want it to be used in the templates by the user or just in the repo for us developers?

**Comment by hmgaudecker** [comment]
Cool! Please enable it for users who opt for "add_python_code_formatter_to_project" and set up git.

I suppose this scans the directories for *.py - patterns. Would it be possible to include "wscript" as well?

**Comment by julienschat** [comment]
I put the pre-commit setup into the post-hooks. It will be installed when both "add_python_code_formatter_to_project" and "set_up_git" are opted for. (f5ce6ae709107bddf5db37f1430dfa281bc376ac)

**Comment by julienschat** [comment]
> All you need is a [config file](https://github.com/tobiasraabe/cookiecutter-research-template/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/.pre-commit-config.yaml) and you need to install all packages under `repo: local`. Then, type `pre-commit install` and the hooks will run before each commit. The posted config runs black, black on documentation, flake8 and doc8. Furthermore, isort is replaced with reorder_python_imports which is simpler and easier to implement.

Hi Tobias, I had to search for a long time for a small bug, that stopped flake8 and black to work properly. I think you excluded all files NOT starting with a dot by mistake, so I changed the exclude regex in the [config file](https://github.com/hmgaudecker/econ-project-templates/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/.pre-commit-config.yaml)

**Comment by tobiasraabe** [comment]
Yes, I also noticed it and changed it some time ago. Now, I exclude only .mywaflib explicitly, but I will look into your approach as it could be more flexible.

**Comment by hmgaudecker** [comment]
I kind of like the "everything with the dot" approach better, I do not think you would ever want to have a script starting with that.

**Comment by hmgaudecker** [comment]
Before closing this, could you (@raholler) please make sure this works as expected and check for any other useful hooks? UTF8-BOM comes to mind, I got bitten by that one a couple of times. Suggestions welcome.

**Comment by hmgaudecker** [comment]
@raholler: Please add an exception for .pre-commit-config.yaml in the .gitignore file

@tobiasraabe: pre-commit runs like a charm, but ... the commit is not carried out. The only way I got it to carry out the commit (not much time to play around), was to uninstall pre-commit temporarily. I installed it as part of the miniconda environment. Did you ever see this? The first few google results were unhelpful...

**Comment by tobiasraabe** [comment]
I am not sure what you mean. The commit is always aborted if one check fails. E.g. if black reformats your files, the commit is aborted and you have to commit again even if the files are now valid. But that is probably not what you mean. I never had problems with pre-commit related to miniconda.

Also can you confirm that doc8 is always skipped because no files were passed to the exec? I thought the list of types is evaluated with OR but maybe it is AND, but no file is markdown and rst.

**Comment by hmgaudecker** [comment]
Ah, 

    In case any of the defined hooks catches an error, the commit is aborted.

I did not see that. Had a bunch of files I moved over and the output was so long that I did not dig through everything. Thought there were only warnings, though, but maybe some file had a syntax error.

Is there any way to override that setting on a per-commit basis? Or anything less invasive than uninstalling pre-commit altogether?

**Comment by tobiasraabe** [comment]
Maybe ``git commit --no-verify``.

**Comment by hmgaudecker** [comment]
Thanks! Will try that out!

**Comment by raholler** [comment]
Here is a list of  [https://pre-commit.com/hooks.html]( pre-commit hooks)

I would suggest:

1. check-byte-order-marker - Forbid files which have a UTF-8 byte-order marker 
2. check-added-large-files - Prevent giant files from being committed (happened to me once accidentally)
3. check-merge-conflict - Check for files that contain merge conflict strings. 
4. check-json - This hook checks json files for parseable syntax.
5. pretty-format-json - This hook sets a standard for formatting JSON files. 
6. trailing-whitespace - This hook trims trailing whitespace. (or is this alrady covered by flake8?)

Maybe:
- check-yaml - This hook checks yaml files for parseable syntax. 
- check-ast - Simply check whether the files parse as valid python. 
- fix-encoding-pragma - Add # -*- coding: utf-8 -*- to the top of python files

**Comment by hmgaudecker** [comment]
Sounds all good, except for the two last maybes (Syntax is checked implicitly by black, utf8 is redundant for python 3).

If 6. only is about python files and does not touch docstrings, it would be redundant indeed.

For 2. we should definitely find out how to override things. Sometimes I do want to add data.

Thanks!

**Comment by raholler** [comment]
Regarding 2., you can exclude certain files in the .pre-commit-config.yaml from this hook. However, this may not be very convenient and too complicated for the casual user. Alternatively, it should be possible to customize the hook (see [https://github.com/pre-commit/pre-commit-hooks/blob/master/pre_commit_hooks/check_added_large_files.py](url)). However, this may need too much effort for only a small convinience hook. What do you think?

**Comment by hmgaudecker** [comment]
I like the possibility to exclude things. Is it possible to print something out on that if the commit fails because of it? 

Btw, we should check that black likes the entire template...

**Comment by raholler** [comment]
>  black likes the entire template

What do you mean?

> Is it possible to print something out on that if the commit fails because of it?

It should say that the commit fails because it failed this check. Do you mean that or do you have something else in mind?

**Comment by hmgaudecker** [comment]
> > black likes the entire template
> 
> What do you mean?

That if you go through the sequence like:

1. Set up a project
2. Set up and activate all pre-commit hooks
3. Commit 

The last step works immediately. Usually that would be hidden by the fact that the initial commit is done before the second step and if people change a file later, the commit might fail because the template did not adhere to the highest standards :-)

That will be a bit annoying to achieve because of the cookiecutter stuff in the raw version, but should be doable.

> > Is it possible to print something out on that if the commit fails because of it?
> 
> It should say that the commit fails because it failed this check. Do you mean that or do you have something else in mind?

Basically, ideally you would want the solutions along with the failure message. I.e.,

- if you added the file on purpose, add an exception to ... in .pre-commit-config.yaml
- if you added the file by mistake, type git reset and add the file to .gitignore
